### PR TITLE
Controller reconciliation rework - part 1 (ServiceInstance)

### PR DESCRIPTION
**Pull request by luksa**
Addresses #1639 and #1363

### CODE REVIEW

File: pkg/controller/controller_instance.go [hunk 1]
```diff
@@ -747,19 +754,20 @@ func (c *controller) processServiceInstancePollingFailureRetryTimeout(instance *
 
 // resolveReferences checks to see if ClusterServiceClassRef and/or ClusterServicePlanRef are
 // nil and if so, will resolve the references and update the instance.
+// If references don't need to be resolved, the method returns false as its 2nd return value
```
Comment (nilebox) [inline pkg/controller/controller_instance.go]: I think this comment is a bit misleading.
For example, we return `false` when `ClusterServiceClassRef` is missing, but there is an error resolving it.
Can we rephrase it to "If references needed to be resolved, and the instance status was successfully updated, the method returns 'true'"?

File: pkg/controller/controller_instance.go [hunk 2]
```diff
@@ -768,16 +776,17 @@ func (c *controller) resolveReferences(instance *v1beta1.ServiceInstance) (*v1be
...
 
 		instance, err = c.resolveClusterServicePlanRef(instance, sc.Spec.ClusterServiceBrokerName)
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
 	}
-	return c.updateServiceInstanceReferences(instance)
+	updatedInstance, err := c.updateServiceInstanceReferences(instance)
+	return updatedInstance, true, err
```
Comment (nilebox) [inline pkg/controller/controller_instance.go]: Shall we return true only if the update has succeeded? i.e.
```go
return updatedInstance, err == nil, err
```

Comment (luksa) [reply pkg/controller/controller_instance.go]: Yes, absolutely. Missed that because I introduced the bool at the last minute (previously, I had the method simply return `nil, nil` to signal that it hadn't modified the instance; afters speaking to Paul about that, I introduced the bool to make the method more explicit).

File: pkg/controller/controller_instance.go [hunk 3]
```diff
@@ -333,6 +335,8 @@ func (c *controller) reconcileServiceInstanceAdd(instance *v1beta1.ServiceInstan
 			// over with a fresh view of the instance.
 			return err
 		}
+		// recordStartOfServiceInstanceOperation has updated the instance, so we need to continue in the next iteration
+		return nil
```
Comment (staebler) [inline pkg/controller/controller_instance.go]: This is going to open the window more for #1792. I don't suggest that we do anything about the issue for this PR. In the future, we'll want to check the in-progress properties to make sure that they haven't changed as well.

File: pkg/controller/controller_instance.go [hunk 4]
```diff
@@ -798,37 +805,39 @@ func (c *controller) processServiceInstancePollingFailureRetryTimeout(instance *
 
 // resolveReferences checks to see if ClusterServiceClassRef and/or ClusterServicePlanRef are
 // nil and if so, will resolve the references and update the instance.
+// If references needed to be resolved, and the instance status was successfully updated, the method returns true as
+// its 2nd return value
 // If either can not be resolved, returns an error and sets the InstanceCondition
 // with the appropriate error message.
-func (c *controller) resolveReferences(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
+func (c *controller) resolveReferences(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, bool, error) {
```
[the remark this answers is not in the document]
Comment (luksa) [reply pkg/controller/controller_instance.go]: Done.

**Comment by luksa** [comment]
I've now also fixed the `reconcileServiceInstanceUpdate()` and `reconcileServiceInstanceDelete()` functions. In other words, the controller will now properly perform the reconciliation for all three ServiceInstance events (add, update & delete). See https://github.com/kubernetes-incubator/service-catalog/pull/1779/commits/832a1b4f89f16adaf948583fe7d458b5f9dfe7c0 for those changes.

I'll finish implementing the changes for ServiceBindings tomorrow and submit them in a separate PR.
